keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2043 @@
+}
+
+// DriverSchedulingStatus contains information provided by the resource driver
+// while handling delayed allocation.
+type DriverSchedulingStatus struct {
+	// Only nodes matching the selector will be considered by the scheduler
+	// when trying to find a Node that fits a Pod. A resource driver can
+	// set this immediately when a ResourceClaim gets created and, for
+	// example, provide a static selector that uses labels.
+	//
+	// Setting this field is optional. If nil, all nodes are candidates.
+	SuitableNodes *core.NodeSelector
+
+	// A change of the PotentialNodes field triggers a check in the driver
+	// on which of those nodes the resource might be made available. It
+	// then excludes nodes by listing those where that is not the case in
+	// UnsuitableNodes.
+	//
+	// Nodes listed here will be ignored by the scheduler when selecting a
+	// node for a Pod. All other nodes are potential candidates, either
+	// because no information is available yet or because allocation might
+	// succeed.
+	//
+	// This can change, so the driver must refresh this information
+	// periodically and/or after changing resource allocation for some
+	// other ResourceClaim until a node gets selected by the scheduler.
+	UnsuitableNodes []string
+}
+
+// AllocationResult contains attributed of an allocated resource.
+type AllocationResult struct {
+	// Attributes contains arbitrary data returned by the driver after a
+	// successful allocation.  This data is passed to the driver for all
+	// operations involving the allocated resource. This is opaque for
+	// Kubernetes.  Driver documentation may explain to users how to
+	// interpret this data if needed.
+	//
+	// The attributes must be sufficient to deallocate the resource because
+	// the ResourceClass might not be available anymore when deallocation
+	// starts.
+	Attributes map[string]string
+
+	// This field will get set by the resource driver after it has
+	// allocated the resource driver to inform the scheduler where it can
+	// schedule Pods using the ResourceClaim.
+	//
+	// A resource driver may already set this before the resource is
+	// allocated. The scheduler will then check this field in addition to
+	// UnsuitableNodes to filter out nodes where the resource cannot be
+	// allocated.
+	//
+	// Setting this field is optional. If nil, the resource is available
+	// everywhere.
+	AvailableOnNodes *core.NodeSelector
+
+	// UserLimit determines how many entities are allowed to use this
+	// resource at the same time. The default is 1. -1 enables the usage by
+	// an unlimited number of users. Individual containers in a pod are not
+	// counted as users, only the Pod is.
+	UserLimit int
+}
+
+type PodSpec {
+   ...
+   // ResourceClaims defines which ResourceClaims must be allocated
+   // and reserved before the Pod is allowed to start. The resources
+   // will be made available to those containers which reference them
+   // by name.
+   ResourceClaims []PodResourceClaim
```

Comment:
What's the limit on this?  16?

Reply:
This is widely regarded as a mistake and not to be emulated :)

Please pick a generous-but-not-too-generous number and when a user screams about it, we can discuss what exactly they are doing :)  How about 32?